### PR TITLE
Add FlaxWhisperForAudioClassification model

### LINKED ISSUE #21779: Add Flax Whisper for audio classification
### Feature request

The PR https://github.com/huggingface/transformers/pull/21754 adds the PyTorch version of WhisperForAudioClassification. It would be great to add the Flax equivalent for cross-library equivalence ♻️

### Motivation

Whisper is an encoder-decoder model for speech recognition. However, we can repurpose the model for other speech tasks, such as audio classification.

Audio classification is the task of mapping from an input speech sequence to a single class prediction. For more details, refer to the task page on the Hub: https://huggingface.co/tasks/audio-classification

For audio classification, we only require a single model output. Thus, we do not need the auto-regressive generation capacities of the Whisper decoder (which is used to generate a sequence of text tokens during speech recognition). Instead, we can just use the Whisper encoder to get hidden states, and add a classification head on top to make class label predictions.

This is analogous to using a Wav2Vec2 model for audio classification: the Wav2Vec2 encoder is used to get hidden states, and a classification head added on top to make class label predictions.

The PR https://github.com/huggingface/transformers/pull/21754 adds the PyTorch version of WhisperForAudioClassification. It required adding a projection layer and classification layer on top of the WhisperEncoder. For more details, refer directly to the pull request.

It would be great to add the Flax equivalent of this model for cross-framework support.

The most difficult part of this PR will be getting the model tester to work. You can see from the PyTorch PR that we require a standalone tester for the audio classification model. This is because the original Whisper model is an encoder-decoder model, but the audio classification model is an encoder-only model. Thus, we require different testing logic.

### Your contribution

Opening this one up to the community! This will be quite a fun JAX/Flax PR! 🚀

If you're interested in tackling this, free to drop a comment in this thread and open a PR when you're ready. More than happy to answer any questions / queries about this integration!

### PR DESCRIPTION
**Pull request by raghavanone**
Fixes #21779

### CODE REVIEW

**Comment by sgugger** [comment]
cc @sanchit-gandhi

**Comment by sgugger** [comment]
The test failures are appearing on this one. Let's fix them and re-merge!

**Comment by sanchit-gandhi** [comment]
We need to make two changes following the updates in #22954! First, we need to assign the attribute `gradient_checkpointing` to the class `FlaxWhisperForAudioClassificationModule`, similar to what we do for `FlaxWhisperForConditionalGeneration`:
https://github.com/huggingface/transformers/blob/a5741d7cb59f8a81d1f5fc7a6b106056d34f9969/src/transformers/models/whisper/modeling_flax_whisper.py#L1176
We then need to forward `self.gradient_checkpointing` to the encoder:
```diff
-         self.encoder = FlaxWhisperEncoder(config=self.config, dtype=self.dtype)
+        self.encoder = FlaxWhisperEncoder(config=self.config, dtype=self.dtype, gradient_checkpointing=self.gradient_checkpointing)
```
This will facilitate gradient checkpointing for the module!

**Comment by raghavanone** [comment]
@sgugger @sanchit-gandhi Done, all tests pass !

**Review by sgugger — APPROVED** [review]
Thnaks a lot!

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 1]
```diff
@@ -1512,6 +1512,7 @@ def update_inputs_for_generation(self, model_outputs, model_kwargs):
 class FlaxWhisperForAudioClassificationModule(nn.Module):
     config: WhisperConfig
     dtype: jnp.dtype = jnp.float32
+    gradient_checkpointing: bool = False
 
     def setup(self) -> None:
         self.encoder = FlaxWhisperEncoder(config=self.config, dtype=self.dtype)
```
[the remark this answers is not in the document]
Comment (raghavanone) [reply src/transformers/models/whisper/modeling_flax_whisper.py:1518]: Yes , will do

File: tests/models/whisper/test_modeling_whisper.py [hunk 2]
```diff
@@ -1430,7 +1430,7 @@ def __init__(
         hidden_dropout_prob=0.1,
         attention_probs_dropout_prob=0.1,
         max_position_embeddings=20,
-        max_source_positions=30,
+        max_source_positions=1500,
```
Comment (ydshieh) [inline tests/models/whisper/test_modeling_whisper.py:1433]: Hi @raghavanone @sanchit-gandhi 

I am wondering why we increase these values a lot ...?

Comment (sanchit-gandhi) [reply tests/models/whisper/test_modeling_whisper.py:1433]: Whisper works intrinsically on a sequence length of 30s inputs (which corresponds to 1500 log mel spectrogram frames)

We could use a shorter context window (i.e. 30s -> 15s), we just need to reduce the feature extractor window length

Comment (sanchit-gandhi) [reply tests/models/whisper/test_modeling_whisper.py:1433]: Fixed in https://github.com/huggingface/transformers/pull/24105
